pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py

```diff
@@ -0,0 +1,529 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import defaultdict
+from copy import deepcopy
+from enum import Enum
+from typing import Union, Tuple, Any, Mapping
+
+from pytorch_lightning.core.step_result import Result
+
+
+# used to map boolean to right LoggerStage values
+class FrozenDict(dict):
+    def __init__(self, *args, **kwargs):
+        self._hash = None
+        super(FrozenDict, self).__init__(*args, **kwargs)
+
+    def __hash__(self):
+        if self._hash is None:
+            self._hash = hash(tuple(sorted(self.items())))  # iteritems() on py2
+        return self._hash
+
+    def _immutable(self, *args, **kws):
+        raise TypeError('cannot change object - object is immutable')
+
+    __setitem__ = _immutable
+    __delitem__ = _immutable
+    pop = _immutable
+    popitem = _immutable
+    clear = _immutable
+    update = _immutable
+    setdefault = _immutable
+
+
+LOOKUP_TABLE = FrozenDict({"1": "test", "0": "validation", "True": "test", "False": "validation"})
+
+
+class LoggerStages(Enum):
+    TRAIN = "train"
+    VAL = "validation"
+    TEST = "test"
+
+
+class StorageType(Enum):
+
+    INSIDE_BATCH_TRAIN_LOOP = "inside_batch_train_loop"
+    ELSE = "else"
```

Comment:
By `ELSE` do you mean outside batch train loop in this case? Can we be more explicit here?

Also `StorageType` is pretty general I think, can we be more specific to the code so it reads better? It might not be possible however, I wasn't able to come up with a suggestion here